api/desecapi/models.py

```diff
@@ -121,9 +121,12 @@ def is_staff(self):
         return self.is_admin
 
     def unlock(self):
-        # self.locked is used by domain.sync_to_pdns(), so call that first
-        for domain in self.domains.all():
-            domain.sync_to_pdns()
+        # Create domains on pdns that were created after the account was locked.
+        # We filter by created__gt=self.locked, published=None should give the
+        # same result as the published timestamp is updated at domain creation.
+        # TODO Should we check somewhere whether the two filters agree?
+        for domain in self.domains.filter(created__gt=self.locked):
```

Comment:
I propose to follow the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle): "Be conservative in what you do, be liberal in what you accept from others". Applying it to the situation at hand, I propose that `unlock()` will unlock all locked domains, regardless of when and why they have been locked. One way to do that could be to wrap each `domain.create_on_pdns()` in a `try`/`except` block and ignore exceptions due to already existing domains.

Reply:
I agree with the principle, but I don't think it applies to the case at hand: If we determine that there is an unpublished domain from the time before the account lock, then that's not a user input to the unlock action. In other words, it is an unexpected internal condition and not something "from others" that we may be willing to accept.

I'm leaning towards checking that the internal condition is in a well-defined state, and otherwise do something conservative, including notifying us.

Also, it may happen in the future that we have unpublished domains for some valid use case, and we may not want to publish them automatically when the user account is unlocked.

So, if you don't have a strong opinion on this, I'll `assert` that either filter condition returns the same number of domains (this is not a perfect check).

Reply:
By "from others", I meant the current state of the database. Let it be legacy code that created that, or manual iteration with it. I'm proposing to state the exact behavior of unlocking in a function doc comment and remove the check, but I do not have a strong opinion.

Reply:
I tried to state behavior more clearly now in the comment, see https://github.com/desec-io/desec-stack/pull/148/commits/4f778bce174302273d09f4dcbe2247f4e0e8d16d#diff-d93a995458c0f3a31fb831c4f43b5d27R127

If this is ok for you, we can merge.